pyfive/dataobjects.py

```diff
@@ -659,7 +595,202 @@ def _decode_link_info_msg(data, offset):
+
+        if data_offset == UNDEFINED_ADDRESS:
+            # no storage is backing array, return all zeros
+            return np.zeros(self.shape, dtype=self.dtype)[args]
+
+        if not isinstance(self.dtype, tuple):
+            if self.avoid_mmap:
+                return self._get_selection_from_contiguous(args)
+            else:
+                try:
+                    # return a memory-map to the stored array with copy-on-write
+                    return np.memmap(self.fh, dtype=self.dtype, mode='c',
+                                offset=data_offset, shape=self.shape, order=self.order)[args]
+                except UnsupportedOperation:
+                    return self._get_selection_from_contiguous(args)
+        else:
+            dtype_class = self.dtype[0]
+            if dtype_class == 'REFERENCE':
+                size = self.dtype[1]
+                if size != 8:
+                    raise NotImplementedError('Unsupported Reference type')
+                ref_addresses = np.memmap(
+                    self.fh, dtype=('<u8'), mode='c', offset=data_offset,
+                    shape=self.shape, order=self.order)
+                return np.array([Reference(addr) for addr in ref_addresses])[args]
+            else:
+                raise NotImplementedError('datatype not implemented')
+
+
+
+    def iterchunks(self, sel=None):
+        """ 
+        Iterate over chunks in a chunked dataset. 
+        The optional sel argument is a slice or tuple of slices that defines the region to be used. #FIXME: sel not yet implemented.
+        If not set, the entire dataspace will be used for the iterator.
+        For each chunk within the given region, the iterator yields a tuple of slices that gives the intersection of the given chunk 
+        with the selection area. This can be used to read or write data in that chunk.
+        A TypeError will be raised if the dataset is not chunked.
+        A ValueError will be raised if the selection region is invalid.
+        """
+        if self.chunks is None:
+            raise TypeError('Dataset is not chunked')
+        return self.id._iter_chunks(sel)
+
+    def _get_selection_from_contiguous(self, args=None):
+        """
+        Two options, we either read the entire contiguous array, and pull out
+        the selection (args) from that, or we try and read the contiguous
+        array in pseudo chunks (being the contiguous data which varies on
+        the slowest axis). We only do the latter if a) we have the
+        pseudo_chunking turned on (see init method), b) the array is 
+        multi dimensional, and c) the slabs are big enough to bother
+        (i.e bigger than pseudo_block_size_kib). We use the fact that the
+        storage order is known (HDF writes in the C-order, last axis
+        varying fastest).
+        """
+
+        # don't want to be doing this if we are actually chunked!
+        assert self.chunks is None
```

Comment:
come on Bryan, prettier ways are available: raise an exception or pass a debug with an empty return, not `assert` :grin: